### PR TITLE
Query: Fix data is undefined on error

### PR DESCRIPTION
**Pull request by TLadd**
If the query fails on the first result, the query result data was being returned as undefined
Now this case will result in data being an empty object

Fixes https://github.com/apollographql/react-apollo/issues/1977

### CODE REVIEW

**Comment by hwillson** [comment]
Thanks for working on this @TLadd! I agree with your assessment that the `data` result should be an empty object. Your changes look great, but I'm a little concerned that this is a breaking change. Some people have likely noticed that `data` is `undefined` when there is an error, and are leveraging that fact in their error handling logic. Rolling this change out could affect that since their `!data` checks will now fail. Hmmm - we'll need to think about this a bit more ... 🤔.

**Comment by TLadd** [comment]
Using !data to check for errors would be pretty brittle anyways, since it only happens if there isn't already a previous result. Certainly could be possible that there is code out there depending on it though.

**Comment by alamothe** [comment]
The typings say `data` is either `TData` or `undefined` - never empty object: https://github.com/apollographql/react-apollo/blob/3985b47f086f38f891099732842ca2f16c7966c4/src/Query.tsx#L97

So which one is it?

**Comment by TLadd** [comment]
Thanks for getting it across the finish line!

File: src/Query.tsx [hunk 1]
```diff
@@ -94,7 +94,7 @@ export interface QueryResult<TData = any, TVariables = OperationVariables>
   // I'm aware of. So intead we enforce checking for data
   // like so result.data!.user. This tells TS to use TData
   // XXX is there a better way to do this?
-  data: TData | undefined;
+  data: TData | {};
```
Comment (derekclair) [inline src/Query.tsx]: 🥇

**Comment by danilobuerger** [comment]
@hwillson @TLadd this should be reverted, see #2424 . This PR promised:
```
instead of `data && data.user` you can just check `data.user`
```
instead it made things worse. Now a type guard is required to discriminate between `TData` and `{}`.

**Comment by tomitrescak** [comment]
Yes please, revert this change. It makes this VERY difficult to use in Typescript ;(

**Comment by hwillson** [comment]
@danilobuerger @tomitrescak Rolling back and we'll have a new release out this morning - thanks!

**Comment by kallaspriit** [comment]
So what ise the solution? Also the TypeScript example does not seem to work (any more) and it uses the old method of:

examples/typescript/src/Character.tsx
```typescript
if (!data) return <div>no data</div>;
```

**Comment by hwillson** [comment]
@kallaspriit For now we'll roll this back, then we'll come up with another solution afterwards (we really didn't want breaking changes in this release). The examples are in the process of being updated and will be available shortly. We're also going to wire the example app tests into CI, to help keep them healthy.

**Comment by kallaspriit** [comment]
Wrote my comment seconds before seeing your reply. Sounds great :)

**Comment by tomitrescak** [comment]
@fenok that is not true. Those are fundamentally different concepts and would validate very differently. Your example also allows partial objects, while the former allows only complete objects or empty objects

**Comment by fenok** [comment]
Valid point, I was trapped in my partial results world. I also found that the idea of `Partial` has already been discussed and rejected. Sorry for the interruption 😅

**Comment by evenfrost** [comment]
So, what's the current status of it?

**Comment by derekclair** [comment]
I believe it was like Typescript@2.5 or something, they introduced a `unknown` type; which, I believe addresses this issue perfectly! If I could request, [a code-review please](https://github.com/derekclair/react-apollo/commit/b9f1e5b2993bb634d7e83cd96e4f5790dc860da3), or two, on this fix that implements that type. I could really use another set of eyes on it.

**Comment by lucasconstantino** [comment]
@evenfrost it was changed, but it's changed back from the previous change, which actually changed to keep things the same they were before the change, so that we don't really change what was change before.

**Comment by IanVS** [comment]
Thank you so much for clarifying, @lucasconstantino.  I was getting really lost in trying to understand where things stand.  💫

**Pull request merged**